rama-net/src/client/pool.rs

```diff
@@ -0,0 +1,618 @@
+use rama_utils::macros::{impl_deref, nz};
+use std::collections::VecDeque;
+use std::fmt::Debug;
+use std::num::NonZeroU16;
+use std::ops::{Deref, DerefMut};
+use std::pin::Pin;
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+use std::{future::Future, net::SocketAddr};
+use tokio::io::{AsyncRead, AsyncWrite};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+use tokio::time::timeout;
+
+#[expect(dead_code)]
+#[derive(Debug)]
+/// Active slot is able to actively use a connection to make requests.
+/// They are used to track 'active' connections inside the pool
+struct ActiveSlot(OwnedSemaphorePermit);
+
+#[expect(dead_code)]
+#[derive(Debug)]
+/// Pool slot is needed to add a connection to the pool. Poolslots have
+/// a one to one mapping to connections inside the pool, and are used
+/// to track the 'total' connections inside the pool
+struct PoolSlot(OwnedSemaphorePermit);
+
+/// A connection which is stored in a pool. A hash is used to determine
+/// which connections can be used for a request. This hash encodes
+/// all the details that make a connection unique/suitable for a request.
+struct PooledConnection<C, ConnId> {
+    /// Actual raw connection that is stored in pool
+    conn: C,
+    /// ID is not unique but is used to group connections that can be used for the same request
+    id: ConnId,
+    /// Slot this connection takes up in the pool
+    slot: PoolSlot,
+}
+
+impl<C: Debug, ConnId: Debug> Debug for PooledConnection<C, ConnId> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PooledConnection")
+            .field("conn", &self.conn)
+            .field("id", &self.id)
+            .field("slot", &self.slot)
+            .finish()
+    }
+}
+
+/// [`LeasedConnection`] is a connection that is temporarily leased from
+/// a pool and that will be returned to the pool once dropped if user didn't
+/// take ownership of the connection `C` with [`LeasedConnection::conn_to_owned()`].
+/// [`LeasedConnection`]s are considered active pool connections until dropped or
+/// ownership is taken of the internal connection.
+pub struct LeasedConnection<C, ConnId> {
+    pooled_conn: Option<PooledConnection<C, ConnId>>,
+    /// Weak reference to pool so we can return connections on drop
+    pool: Weak<PoolInner<C, ConnId>>,
+    _slot: ActiveSlot,
+}
+
+impl<C: Debug, ConnId: Debug> Debug for LeasedConnection<C, ConnId> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("LeasedConnection")
+            .field("pooled_conn", &self.pooled_conn)
+            .field("pool", &self.pool)
+            .field("_slot", &self._slot)
+            .finish()
+    }
+}
+
+impl<C, ConnId> LeasedConnection<C, ConnId> {
+    /// Take ownership of the internal connection. This will remove it from the pool.
+    pub fn take(mut self) -> C {
+        let conn = self.pooled_conn.take().expect("only None after drop").conn;
```

Comment:
it is pretty unconventional to assign it to a local variable to immediately return it. I also do not see it simplifying any logic here